### PR TITLE
#208: improving how let statements are resolved using stack and name …

### PR DESCRIPTION
**Pull request by c-schuler**
…mangling

Also cleaned up the build for release.

### CODE REVIEW

File: cql-engine/src/main/java/org/opencds/cqf/cql/elm/execution/ExpressionDefEvaluator.java [hunk 1]
```diff
@@ -14,10 +14,13 @@ protected Object internalEvaluate(Context context) {
         }
         try {
             if (context.isExpressionCachingEnabled() && context.isExpressionInCache(this.getName())) {
+                context.setCurrentExpression(this.getName());
                 return context.getExpressionResultFromCache(this.getName());
             }
 
+            context.setCurrentExpression(this.getName());
             Object result = this.getExpression().evaluate(context);
```
Comment (brynrhodes) [inline cql-engine/src/main/java/org/opencds/cqf/cql/elm/execution/ExpressionDefEvaluator.java]: This should be protected with a try..finally, yes?

File: cql-engine/src/main/java/org/opencds/cqf/cql/elm/execution/QueryEvaluator.java [hunk 2]
```diff
@@ -256,8 +256,6 @@ protected Object internalEvaluate(Context context) {
             return null;
         }
 
-        context.clearLetExpressions();
-
```
Comment (brynrhodes) [inline cql-engine/src/main/java/org/opencds/cqf/cql/elm/execution/QueryEvaluator.java]: We still need to clear let expressions, right?

File: cql-engine/src/main/java/org/opencds/cqf/cql/elm/execution/ExpressionDefEvaluator.java (hunk 1 above)
Comment (c-schuler) [reply cql-engine/src/main/java/org/opencds/cqf/cql/elm/execution/ExpressionDefEvaluator.java]: Isn't it?

File: cql-engine/src/main/java/org/opencds/cqf/cql/elm/execution/QueryEvaluator.java (hunk 2 above)
Comment (c-schuler) [reply cql-engine/src/main/java/org/opencds/cqf/cql/elm/execution/QueryEvaluator.java]: No, they do not need to be cleared the way they are saved now.

File: cql-engine/src/main/java/org/opencds/cqf/cql/elm/execution/ExpressionDefEvaluator.java (hunk 1 above)
Comment (brynrhodes) [reply cql-engine/src/main/java/org/opencds/cqf/cql/elm/execution/ExpressionDefEvaluator.java]: No. It's true that it's inside a larger try..finally, but that is protecting the evaluation context. In line 21, we are pushing the current expression onto a stack. In line 23, we are popping it. If an error occurs in line 22, line 23 will never execute, and we'll be leaving the stack imbalanced.

**Comment by c-schuler** [comment]
Closing due to a requested change of approach, which renders the work here unusable.

**Pull request closed without merging**